### DISCUSSION TITLE
Awaitable Server Start()

### DISCUSSION THREAD

**Discussion by seikosantana** [scottoffen/grapevine#80] [Q&A] [answered]
Instead of calling `Start()` and then holding the process by `Console.ReadLine()`, can we `await` the `Start()` either as a `Task` or a `Thread`? I will be running a Console app as a systemd service and since standard input stream will be connected to `/dev/null` then `Console.ReadLine()` might not work properly to hold the process. Yes it is possible to create a unit that allocates a TTY for the console app, but I prefer to wait the execution on main thread.

If `IRestServer.Start()` returns task or thread, it will be possible to `await` or `Join()`. But since it is not, do we have another alternatives so we don't rely on waiting for input?

**Comment by scottoffen** [comment, marked as answer]
There are two alternatives to `server.Start()` available via extension methods on `IRestServer` for use with Console applications.

```csharp
// This option will hold the process until the server stops
server.Run();

// This option will hold the process until the cancellation token is canceled
server.Run(CancelationToken token);
```

> If you do use one of these options, you do not need to explicitly call `server.Start()`.
